### PR TITLE
Fix chrome_android to work with emulator

### PR DESCRIPTION
**Pull request by galmacky**
This is a hacky workaround to make it run against Chrome on Android.

### CODE REVIEW

File: tools/wptrunner/wptrunner/browsers/chrome_android.py [hunk 1]
```diff
@@ -50,7 +50,10 @@ def executor_kwargs(test_type, server_config, cache_manager, run_info_data,
     capabilities = dict(DesiredCapabilities.CHROME.items())
     capabilities["chromeOptions"] = {}
     # required to start on mobile
-    capabilities["chromeOptions"]["androidPackage"] = "com.google.android.apps.chrome"
+    # Workaround: package name should reflect channel correctly.
+    capabilities["chromeOptions"]["androidPackage"] = "com.android.chrome"
+    # Workaround: ANDROID_SERIAL is ignored in webdriver.
+    capabilities["chromeOptions"]["androidDeviceSerial"] = "emulator-5554"
```
Comment (Hexcles) [inline tools/wptrunner/wptrunner/browsers/chrome_android.py]: What is `emulator-5554`? Is this just happen to be the device ID for your emulator that you were running? I don't remember having to set this last time. Is it only required when there are multiple devices connected?

File: tools/wptrunner/wptrunner/executors/executorselenium.py [hunk 2]
```diff
@@ -187,6 +187,9 @@ def send_message(self, message_type, status, message=None):
             obj["message"] = str(message)
         self.webdriver.execute_script("window.postMessage(%s, '*')" % json.dumps(obj))
 
+def _isAndroidCapabilities(capabilities):
```
Comment (Hexcles) [inline tools/wptrunner/wptrunner/executors/executorselenium.py]: I think in general reftests don't (or at least didn't) really work on Android. Last time I tried ChromeDriver wasn't able to take screenshots from Chrome for Android. Besides, even if we are able to take screenshots, most reftests are written with the assumption that the screen is bigger than a certain size (800x600) and on small mobile screens important parts of the tests might be out of the viewport.

File: tools/wptrunner/wptrunner/browsers/chrome_android.py (hunk 1 above)
Comment (galmacky) [reply tools/wptrunner/wptrunner/browsers/chrome_android.py]: You're correct. I'll remove this part as this only concerns multiple device scenarios. chromedriver should use ANDROID_SERIAL to detect the current device, which it doesn't at the moment, and it should be fixed in chromedriver instead.

File: tools/wptrunner/wptrunner/executors/executorselenium.py (hunk 2 above)
Comment (galmacky) [reply tools/wptrunner/wptrunner/executors/executorselenium.py]: Hmm.. How about running the tests on a certain model of tablet using emulator?

BTW, could you give me an example of reftest?

Comment (jgraham) [reply tools/wptrunner/wptrunner/executors/executorselenium.py]: FWIW Gecko has some magic where it ends up taking a screenshot on a viewport that's not fixed to the actual dimensions of the device.

Also note that reftests tend to require 1dppx to work correctly, so it's highly likely that modern phones do have enough physical pixels, but are usually operating at at least 2dppx.

File: tools/wptrunner/wptrunner/browsers/chrome_android.py (hunk 1 above)
Comment (Hexcles) [reply tools/wptrunner/wptrunner/browsers/chrome_android.py]: https://cs.chromium.org/chromium/src/chrome/test/chromedriver/chrome_launcher.cc?l=591&rcl=bd218f82c33a10cdf226840ce28c597080ceb558
IIRC, ChromeDriver simply uses a bunch of `adb` commands underneath for finding devices etc., so it should be an easy fix.

File: tools/wptrunner/wptrunner/executors/executorselenium.py (hunk 2 above)
Comment (Hexcles) [reply tools/wptrunner/wptrunner/executors/executorselenium.py]: Perhaps the most basic reftest: `infrastructure/reftest/green-ref.html`

File: tools/wptrunner/wptrunner/browsers/chrome_android.py (hunk 1 above)
Comment (galmacky) [reply tools/wptrunner/wptrunner/browsers/chrome_android.py]: Thanks for the suggestion. Landing this CL: https://chromium-review.googlesource.com/c/chromium/src/+/1652556

Comment (galmacky) [reply tools/wptrunner/wptrunner/browsers/chrome_android.py]: Removed now.

File: tools/wptrunner/wptrunner/executors/executorselenium.py (hunk 2 above)
Comment (galmacky) [reply tools/wptrunner/wptrunner/executors/executorselenium.py]: I don't think I can fix reftests in this PR. The goal is to make this minimal viable implementation.
PTAL

File: tools/wptrunner/wptrunner/browsers/chrome_android.py [hunk 3]
```diff
@@ -48,18 +48,20 @@ def executor_kwargs(test_type, server_config, cache_manager, run_info_data,
...
-    capabilities["chromeOptions"] = {}
-    # required to start on mobile
-    capabilities["chromeOptions"]["androidPackage"] = "com.google.android.apps.chrome"
+    capabilities["goog:chromeOptions"] = {}
+    # TODO(chrome): browser_channel should be properly supported.
+    package_name = "com.android.chrome"  # stable channel
+    # Required to start on mobile
+    capabilities["goog:chromeOptions"]["androidPackage"] = package_name
+    # Note: from ChromeDriver 75 and onwards only w3c is supported.
+    capabilities["goog:chromeOptions"]["w3c"] = True
```
Comment (Hexcles) [inline tools/wptrunner/wptrunner/browsers/chrome_android.py]: I think this is the default now and there's no need to specify it?

File: tools/wptrunner/wptrunner/browsers/chrome_android.py [hunk 4]
```diff
@@ -86,15 +88,18 @@ def __init__(self, logger, binary, webdriver_binary="chromedriver",
...
     def _adb_run(self, args):
         self.logger.info('adb ' + ' '.join(args))
         subprocess.check_call(['adb'] + args)
 
-    def setup(self):
+    def setup_adb_reverse(self):
         self._adb_run(['wait-for-device'])
         self._adb_run(['forward', '--remove-all'])
         self._adb_run(['reverse', '--remove-all'])
+        # "adb reverse" basically forwards network connection from device to
```
Comment (Hexcles) [inline tools/wptrunner/wptrunner/browsers/chrome_android.py]: Nit:

```suggestion
        # "adb reverse" forwards network connection from device to host.
```

And drop the next line.

File: tools/wptrunner/wptrunner/executors/executorselenium.py (hunk 2 above)
Comment (Hexcles) [reply tools/wptrunner/wptrunner/executors/executorselenium.py]: I think this change is probably hidden a bit too deep for a workaround. We should probably skip reftests using the command line flag (`--test-types wdspec,testharness`, i.e. exclude `reftest`), then `SeleniumRefTestExecutor` won't be exercised and the runner won't try to set the window size.

**Comment by Hexcles** [comment]
Sorry for the delay, @galmacky . Apart from the comments I just left, the rest of the change LGTM.

The unit test failures are unrelated and will likely go away after you push to the branch.

File: tools/wptrunner/wptrunner/browsers/chrome_android.py (hunk 3 above)
Comment (galmacky) [reply tools/wptrunner/wptrunner/browsers/chrome_android.py]: Removed.

File: tools/wptrunner/wptrunner/executors/executorselenium.py (hunk 2 above)
Comment (galmacky) [reply tools/wptrunner/wptrunner/executors/executorselenium.py]: Good idea. Removed for now.

File: tools/wptrunner/wptrunner/browsers/chrome_android.py (hunk 4 above)
Comment (galmacky) [reply tools/wptrunner/wptrunner/browsers/chrome_android.py]: Done.

**Comment by galmacky** [comment]
Sorry for the delay in re-uploading. I wasn't able to prioritize this work. Could you take another look?

**Review by Hexcles — APPROVED** [review]
LGTM

**Pull request merged**